be2-scala/src/main/scala/ch/epfl/pop/json/MessageDataProtocol.scala

```diff
@@ -459,4 +461,149 @@ object MessageDataProtocol extends DefaultJsonProtocol {
+    final private val PARAM_OTHER_ORGANIZER: String = "other_organizer"
+    final private val PARAM_CHALLENGE: String = "challenge"
+
+    override def read(json: JsValue): FederationInit = json.asJsObject().getFields(PARAM_ID, PARAM_SERVER_ADDRESS, PARAM_OTHER_ORGANIZER, PARAM_CHALLENGE) match {
+      case Seq(laoId @ JsString(_), JsString(serverAddress), other_organizer @ JsString(_), challenge @ JsObject(_)) =>
+        FederationInit(
+          laoId.convertTo[Hash],
+          serverAddress,
+          other_organizer.convertTo[PublicKey],
+          challenge.convertTo[Message]
+        )
+      case _ => throw new IllegalArgumentException(s"Can't parse json value $json to a FederationInit object")
+    }
+
+    override def write(obj: FederationInit): JsValue = {
+      var jsObjectContent: ListMap[String, JsValue] = ListMap[String, JsValue](
+        PARAM_OBJECT -> JsString(obj._object.toString),
+        PARAM_ACTION -> JsString(obj.action.toString),
+        PARAM_ID -> obj.laoId.toJson,
+        PARAM_SERVER_ADDRESS -> obj.serverAddress.toJson,
+        PARAM_OTHER_ORGANIZER -> obj.other_organizer.toJson,
+        PARAM_CHALLENGE -> obj.challenge.toJson
+      )
+      JsObject(jsObjectContent)
+    }
+  }
+
+  implicit object FederationRequestChallengeFormat extends JsonFormat[FederationRequestChallenge] {
+    final private val PARAM_TIMESTAMP: String = "timestamp"
+
+    override def read(json: JsValue): FederationRequestChallenge = json.asJsObject().getFields(PARAM_TIMESTAMP) match {
+      case Seq(timestamp @ JsNumber(_)) => FederationRequestChallenge(
+          timestamp.convertTo[Timestamp]
+        )
+      case _ => throw new IllegalArgumentException(s"Can't parse json value $json to a FederationChallengeRequest object")
+    }
+
+    override def write(obj: FederationRequestChallenge): JsValue = {
+      var jsObjectContent: ListMap[String, JsValue] = ListMap[String, JsValue](
+        PARAM_OBJECT -> JsString(obj._object.toString),
+        PARAM_ACTION -> JsString(obj.action.toString),
+        PARAM_TIMESTAMP -> obj.timestamp.toJson
+      )
+      JsObject(jsObjectContent)
+    }
+  }
+
+  implicit object FederationResultFormat extends JsonFormat[FederationResult] {
+    final private val PARAM_STATUS: String = "status"
+    final private val PARAM_REASON: String = "reason"
+    final private val PARAM_ORGANIZER: String = "organizer"
+    final private val PARAM_CHALLENGE_MESSAGE = "challenge_message"
+
+    override def read(json: JsValue): FederationResult =
+      val jsObj = json.asJsObject
+      val status = jsObj.fields.get(PARAM_STATUS).collect {
+        case JsString(s) => s
+      }.getOrElse(throw new IllegalArgumentException(s"Can't parse json value $json to a FederationResult object"))
+
+      val reason = jsObj.fields.get(PARAM_REASON).collect {
+        case JsString(r) => r
+      }
+
+      val organizer = jsObj.fields.get(PARAM_ORGANIZER).flatMap(_.convertTo[Option[PublicKey]])
```

Comment:
nit : can't the same logic as reason be applied here ? 
val organizer = jsObj.fields.get(PARAM_ORGANIZER).collect {
        case key @ JsString(_) => key.convertTo[PublicKey]
    }